src/launch/nss-cache.c

```diff
@@ -114,45 +114,50 @@ static int nss_cache_add(CRBTree *tree_by_name,
-                                          name,
-                                          &parent_by_name);
-        slot_by_id = c_rbtree_find_slot(tree_by_id,
-                                        nss_cache_node_compare_id,
-                                        (void *)(unsigned long)id,
-                                        &parent_by_id);
-
-        if (slot_by_name)
-                node_by_name = NULL;
-        else
-                node_by_name = c_rbnode_entry(parent_by_name, NSSCacheNode, rb_by_name);
+        if (name) {
+                slot_by_name = c_rbtree_find_slot(tree_by_name,
+                                                  nss_cache_node_compare_name,
+                                                  name,
+                                                  &parent_by_name);
 
-        if (slot_by_id)
-                node_by_id = NULL;
-        else
-                node_by_id = c_rbnode_entry(parent_by_id, NSSCacheNode, rb_by_id);
+                if (!slot_by_name)
+                        node_by_name = c_rbnode_entry(parent_by_name, NSSCacheNode, rb_by_name);
+        }
+
+        if (id != -1) {
+                slot_by_id = c_rbtree_find_slot(tree_by_id,
+                                                nss_cache_node_compare_id,
+                                                (void *)(unsigned long)id,
+                                                &parent_by_id);
+                if (!slot_by_id)
+                        node_by_id = c_rbnode_entry(parent_by_id, NSSCacheNode, rb_by_id);
+        }
 
         /*
          * Either the entry is linked by-name *AND* by-id (and it is the *SAME*
-         * entry), or it is not linked at all. However, if only one of both is
-         * known, or if they point to different entries, we will *NOT* cache
-         * the new data. This situation implies a UID/GID conflict and there is
-         * no sane way to handle this, other than bypassing the cache. Worst
-         * case, you will end up calling into NSS all the time.
+         * entry), it is linked by one and the other is invalid, or it is not
+         * linked at all. However, if only one of both is known and valid, or
+         * if they point to different entries, we will *NOT* cache the new data.
+         * This situation implies a UID/GID conflict and there is no sane way to
+         * handle this, other than bypassing the cache. Worst case, you will
+         * end up calling into NSS all the time.
          */
-        if (node_by_name != node_by_id) {
+        if ((node_by_id && node_by_id->name && !node_by_name) ||
+            (node_by_name && node_by_name->id != -1 && !node_by_id) ||
+            (node_by_name && node_by_id && node_by_name != node_by_id)) {
                 *nodep = NULL;
                 return 0;
         }
 
-        if (!node_by_name && !node_by_id) {
+        if (!node_by_name && !node_by_id && (name || id != -1)) {
```

Comment:
This last check is redundant, isn't it? Also, if neither `name` nor `id` is set, this function will not complain but return `NULL` and success. That is weird, isn't it?

Reply:
The check is to deal with the case of passing in invalid both `name` and `id` (which would otherwise be added to the cache). I moved that to an explicit check at the top, and made it a failure.